### DISCUSSION TITLE
Coalesce Type Inference

### DISCUSSION THREAD

**Discussion by navilan** [Gremlinq/ExRam.Gremlinq#127] [Q&A] [answered]
Hi,

I'm trying to write idempotent queries to create edges in cosmos db. Here is a simplified query:

```C#
            var query = graph.E<Liked>(id).Fold()
                             .Coalesce(
                                  __ => __.Unfold(),
                                  _ => _.V<User>(userId).As(userLabel)
                                        .V<Article>(puzzleId).As(articleLabel)
                                        .AddE<Liker>()
                                        .To(userLabel)
                                        .InV<Article>()
                                        .AddE<Liked>().From(userLabel)
                                        .Property("Id",     id)
                                        .Property("OnDate", now)
                              );
```

What I want to accomplish is this:

```groovy
g
    .E('FE9AD68526F641E4-l-20210129T040540-EDE03C94C36783FB')
    .fold()
    .coalesce(
        unfold(),
        __.V('FE9AD68526F641E4').hasLabel('User').as('u')
        .V('20210129T040540-EDE03C94C36783FB').hasLabel('Article').as('p')
        .addE('Liked')
        .from('u')
        .property('id', 'FE9AD68526F641E4-l-20210129T040540-EDE03C94C36783FB')
        .property('OnDate', '01/30/2021 22:05:04')
        .outV()
        .addE('Liker').to('p'))
```

However, I'm having trouble as the compiler cannot infer the type of `Coalesce`. So, I have two questions:

1. Is my approach towards writing idempotent queries to create edges alright? Do you have other suggestions?
2. What's the best way to provide a type parameter to Coalesce. I find it fairly cumbersome having to come up with the type parameter. Do you have any pointers to make that easy?

Thank you so much,
Cheers
Navilan.

**Comment by danielcweber** [comment, marked as answer]
1. Coalesce is the preferred approach for these kind of queries, however, you should be aware that it's not 100% safe for concurrency because of the lack of transactions in CosmosDB. So the moment you create the edge with a specific Id, another process might have done so already.
2. Check out the types of both current arguments to Coalesce. They probably disagree (one of them probably being IEdge.... and the other one being IOutEdge... or IInEdge... Either explicitly specify IEdgeGremlinQueryBase&lt;Liked> as argument to Coalesce, or force a common base interface by calling 'ForceEdge' on both arguments (the latter one will already do I guess).

**Reply by navilan** [reply]
Thank you @danielcweber .
